### ISSUE TITLE
FindTypeMapFor ignores type parameters when src/dest objects passed

### ISSUE THREAD

**Issue by cherrydev** [AutoMapper/AutoMapper#724]
The offending line is line 345 of ConfigurationStore (at commit c986717, at least)

`return FindTypeMapFor(source?.GetType() ?? sourceType, destination?.GetType() ?? destinationType);`

I'm not sure what was actually intended, so I don't have a PR for you, but obviously discarding `sourceType` and `destinationType` when `source` and `destination` are designated is not what's intended, since it means that when calling `Mapper.Map<TSource,TDest>(srcObject)` TSource is ignored and effectively replaced with `srcObject.GetType()`.  This breaks when, for example, you're using EF dynamic proxies and `typeof(TSource) != srcObject.GetType()`.

**Comment by jbogard** [comment]
Part of the intention is to try and simplify things - the rules for determining what TypeMap to use were confusing as all get out.

What is supposed to happen is srcObject.GetType() might not match, but AutoMapper will traverse the parent types until it finds a match. What does the type hierarchy look like for EF proxies? `Mapper.Map<TSource, TDestination>(srcObject)` still enforces that srcObject is an object that can be cast to TSource via the generic type parameter.  You can't do `Mapper.Map<Duck, DuckDto>(giraffe)`

**Comment by cherrydev** [comment]
Well, as I said, I wasn't quite sure what the intention was, but this change leads to a regression that I can't find a workaround for.  #723 has more information on the specifics of the circumstance.  But now, under some circumstances, if I have:
```csharp
abstract class Animal {
…
}

class Duck extends Animal {
…
}

cfg.CreateMap<Animal, AnimalDto>()
  .Include<Duck, DuckDto>
  …
  .ReverseMap();

…

cfg.CreateMap<Duck, DuckDto>().ReverseMap();
```

As of that commit, `Mapper.Map<Animal,AnimalDto>(aDuck)` fails with `Instances of abstract classes cannot be created.` but (as far as I can tell) only when aDuck is a dynamic proxy subclass of Duck.

I'm creating my mapping exactly as described [here](https://github.com/AutoMapper/AutoMapper/wiki/Mapping-inheritance#mapping-configuration-inheritance-in-automapper-20)

**Comment by cherrydev** [comment]
Okay, I've found the problem.  The algorithm that you're using for searching for an appropriate mapping pair in `FindTypeMapFor(TypePair typePair)` is wrong and the results of it are being cached.

Here's our _typeMaps:

src | dest | note
--- | ---|---
Duck | DuckDto | Our target
Animal | AnimalDto | Invalid, abstract
etc…|etc…

When we call `Mapper.Map<Animal, AnimalDto>(aDuck)` when duck is a (dynamic) subclass of Duck GetRelatedTypePairs will return, in this order:

src|dest|note
---|---|---
DuckProxyClassFoo | AnimalDto
DuckProxyClassFoo | object
Duck | AnimalDto
Duck | object
**Animal** | **AnimalDto**| Selected, but invalid
Animal | object
etc… | etc…

So, the mapping that it will find is Animal -> AnimalDto, which isn't valid.  It will then associate DuckProxyClassFoo -> AnimalDto with Animal -> AnimalDto in the cache.

The correct entry in _typeMaps that should be found and cached for DuckProxyClassFoo -> AnimalDto is Duck -> DuckDto.

Correct me if I'm wrong, but the destType that is considered for finding a match is actually irrelevant as long as it is assignable from the typePair.DestinationType.

A possibility for finding a match where srcType and dstType are the types as currently passed to `FindTypeMapFor(Type sourceType, Type destinationType)` :
* sort _typeMaps according to key.SourceType where more derived types are all at the top
* search sorted _typeMaps and find the first key where srcType is assignable to key.SourceType and dstType is assignable *from* key.DestinationType.

src | dest | note
--- | ---|---
Duck | DuckDto | srcType (DuckProxy) assignable to Duck, AnimalDto assignable from DuckDto
Animal | AnimalDto | also matches, but above is selected already
etc…|etc…

I'm currently trying to figure out a sane way to sort classes by hierarchy…

**Comment by cherrydev** [comment]
Sorry, I have the changes ready, but the line endings seem to be inconsistent so any pull request that I make ends up looking terribly messy.  Please take a look at [eeac7b4](https://github.com/cherrydev/AutoMapper/commit/eeac7b4b70d552a02d34704e28e64b12da3457d8) and see if that looks like it will work for you.  It maintains a sorted list of TypePairs, sorted by the hierarchy depth of pair.SourceType.  When finding a mapping at runtime, it finds the first TypePair that's compatible and selects/caches it.

**Comment by TylerCarlson1** [comment]
Another problem is there's no recursion.
And `FindTypeMapFor(TypePair)` LIES and should be private and called something else like FindTypeMapFromCached() which is what it actually does.

line 388 to be replaced with this will fix the problem.
```
var typeMap = _typeMapCache.GetOrAdd(typePair, _ => GetRelatedTypePairs(_).Select(tp => FindTypeMapFor(tp.SourceType, tp.DestinationType)).FirstOrDefault(tm => tm != null));
```

**Comment by TylerCarlson1** [comment]
@cherrydev as for the ordered depth heirarch it has that functionality already.
It's just when you go from `DuckProxyClassFoo->AnimalDto` to `Duck->AnimalDto` it should recursively try FindTypeMap for `Duck->AnimalDto` the same way it would try to map if your mapping source was Duck instead of DuckProxyClassFoo.

And it still keeps with the simplicity of the original implementation.

Again I stress `FindTypeMapFor(TypePair)` LIES and caused this problem.

**Comment by cherrydev** [comment]
That code results in a stack overflow.  Perhaps I misunderstood you and you intended another change that would fix it.

If it recursively tries from `DuckProxyClassFoo->AnimalDto` to `Duck->AnimalDto` it will still never find the valid mapping, because there is no `Duck->AnimalDto` mapping! The valid mapping is actually `Duck->DuckDto` and GetRelatedTypePairs() will never return that because AnimalDto is the most specific class that's ever passed to FindTypeMapFor and it only ever traverses towards the base class.

SourceType and DestinationType are not symmetrical, as your code implies, since it only ever considers them as a pair.

Think about it: The source type is ALWAYS the most derived class (because it's coming from the actual object you're mapping) and will always either be the mapped SourceType or a subclass of it.  On the other hand, the destination class will always either be the mapped DestinationType or a **super**class of it.  If you call `Map<TSource,TDest>()` the mapper that's selected can NEVER be a base class of TDest because it cannot assign the resulting object to TDest.

As long as the code is calling GetRelatedTypePairs on the original TypePair and trying to find, from the list of TypePairs, a matching TypePair in _typeMaps, I can't see how it will ever be correct, because it will never work in the case that TDest is a base class of the target TypeMap's DestinationType that we're looking for, at least not according to the documentation on how to write your mappings.

In any case, the commit that I pointed you at works, unless there's some sort of special handling that's needed for dealing with mapped interfaces (in which case I think it'll still work because interfaces will end up at the bottom of _orderedTypePairs since they have no BaseType, but I haven't tested it).  I'm absolutely sure that there are many other ways to make it work as well, but I don't think it can work with the strategy you seem to be using.

Regardless, the current version breaks inheritance mapping and I wonder if you might consider reverting the commit that broke it if my patch is insufficient and you don't have a another solution already.

Thanks for being really responsive!

**Comment by cherrydev** [comment]
Wait, I get where you're going with this:

There's no `Duck->AnimalDto` mapping in _typeMaps, but there IS one in the *cache* already that *should* point to the `Duck->DuckDto` mapping because the cache entry was put there on Seal(), right?

**Comment by cherrydev** [comment]
Okay, here:
```csharp
var typeMap = _typeMapCache.GetOrAdd(typePair, _ => GetRelatedTypePairs(_).Select(tp => _typeMapCache.ContainsKey(tp) ? _typeMapCache[tp] : null).FirstOrDefault(tm => tm != null));
```

It doesn't need to be recursive at all!  Is this correct?  It solves my needs (at least after calling Seal()).

**Comment by jbogard** [comment]
I don't think it should be since the code that originally finds the type
map (non-cached version) is recursive. That GetRelatedTypePairs method that
is.

Also, if you remove the TypePair caching on Seal, does that help? That was
put in on a whim.

What I'm trying to do is reduce runtime figuring out of map plans and
figure out more ahead of time.

**Comment by cherrydev** [comment]
I think that if you're fine with Seal() becoming required for correct operation, the caching is good.  If you remove the caching, the above code won't work at all.  I think you'll have to make an explicit decision about `Seal()`, though: If `Seal()` will, from now on, be required for AutoMapper to behave properly, I think it should fail fast if you neglect to call it, rather than just failing under only some circumstances.
Another option is that Seal() gets called explicitly the first time any mapping is performed, but then you'll have to muck around with thread safety.

So, I didn't quite understand from your last message: Is the new fix I'm proposing what you had in mind?  There's a PR for it, if so.

**Comment by TylerCarlson1** [comment]
@cherrydev Sorry that's what I meant to achieve is in 930aceb.  Recursively checks cached list.  Reading the cache vs the type maps list in order to find the inheritance pair mapping.  Also since you are looking in cache probably don't need to do get or update just FirstOrDefault().

**Comment by cherrydev** [comment]
Well, if we leave the cache as GetOrAdd, then we will end up with an item in the cache with the key `DuckProxyClassFoo->AnimalDto` that will point to `Duck->DuckDto`.  That's exactly what we want, isn't it?

**Comment by TylerCarlson1** [comment]
Yes you are right.  I spoke too soon.

**Comment by jbogard** [comment]
Well hopefully this simplifies things. The TypeMap resolution code before
was....crazy. Rather than try and refactor it, I just ditched it and tried
and algorithm I thought would work. All tests passed but clearly there were
still edge cases.

**Comment by cherrydev** [comment]
Ya, even though `FindTypeMapFor(TypePair)` never actually worked at all when called during an inherited mapping, it didn't have to because SOME of the keys (`Duck->AnimalDto`) were already in the cache from `Seal()`.  It broke as soon as the cache missed (because `DuckProxyClassFoo->AnimalDto` wasn't in the cache).  That should be easy to add a test for now that it's a known case.

**Comment by TylerCarlson1** [comment]
Double checked unit tests and it looks like you need to run both `FindTypeMapFor(TypePair)` and `tp => _typeMapCache.ContainsKey(tp) ? _typeMapCache[tp] : null`, else you will have to require `Mapper.Initialize` for all mappings, and not just inheritance.
Half the tests failed because they didn't have `Mapper.Initialize` and thus weren't in the _typeMapCache.

**Comment by cherrydev** [comment]
Well, that's part of what I was saying:  Is `Initialize`/`Seal()` required or not?  If it's not required, there should not be cases where it fails if you don't call it.  If it is required, I think it needs to fail fast.

**Comment by cherrydev** [comment]
Okay, https://github.com/AutoMapper/AutoMapper/pull/727 now everything passes.

**Comment by jbogard** [comment]
So on the topic of naming things:

FindTypeMapFor - finds explicitly defined type maps (those in the `_typeMaps` collection)
ResolveTypeMap - goes through the cache (I'm going to rename that to something like 'typeMapPlan' or something).

What do you think @TylerCarlson1 ?

**Comment by TylerCarlson1** [comment]
That sounds better than what you have now.
I would suggest possibly putting Try before each one because they aren't guaranteed to get a result, but that's me being very picky about the names against their intent.
Maybe keep _typeMapCache the same name and change_typeMaps to _definedTypeMaps or _userDefinedTypeMaps, so they are different.
Of if you do change _typeMapCache to _typeMapPlan I would still suggest changing _typeMaps name.  _typeMaps isn't very clear name, especially now that you have a another TypeMap list.

**Comment by jbogard** [comment]
Hmmm...I'm careful with TryXyz because that's a specific pattern with .NET - the Try-Parse pattern https://msdn.microsoft.com/en-us/library/ms229009%28v=vs.110%29.aspx but I like the names you got for the fields.

So I'm adding overloads, FindTypeMapFor just looks up user-defined type maps, ResolveTypeMap will use type resolution rules. That should clean things up a bit!

**Comment by TylerCarlson1** [comment]
So ResolveTypeMap will have every overload FindTypeMapFor has now and both will be in IConfigurationStore with all the overloads?

**Comment by jbogard** [comment]
Not exactly - I'm moving some FindTypeMapFor over to ResolveTypeMap. Here's what I'm thinking:

```
TypeMap FindTypeMapFor(Type sourceType, Type destinationType);
TypeMap FindTypeMapFor(TypePair typePair);

TypeMap ResolveTypeMap(object source, object destination, Type sourceType, Type destinationType);
TypeMap ResolveTypeMap(Type sourceType, Type destinationType);
TypeMap ResolveTypeMap(TypePair typePair);
TypeMap ResolveTypeMap(ResolutionResult resolutionResult, Type destinationType);
```

And make it explicit and obvious that the resovler versions take additional information (source/destination values) and the FindTypeMapFor only access the user-defined type map list for those explicit types you've passed in.

**Comment by TylerCarlson1** [comment]
Having different set of overloads will help.

I was wondering do you need FindTypeMapFor anymore at this point?
From what I saw was FindTypeMapFor that just looks at _typeMap was only being used internally and all other implementations of FindTypeMapFor were using the ResolveTypeMap implementation.

Is there any reason a user would want to see a type map is explicitly defined vs resolved?

**Comment by jbogard** [comment]
Look at the latest push, that might clear it up. It also resolved the case
where people query the configuration

**Comment by TylerCarlson1** [comment]
Can't view code in VS, don't have 2015 on this machine.  I can't tell for sure but did you replace every usage of FindTypeMapFor with ResolveTypeMap?  If so then do you really need FindTypeMapFor?

**Comment by TylerCarlson1** [comment]
Sorry every usage of FindTypeMapFor that's not in ConfigurationStore class is what I meant to say.  Meaning it could technically be a private method.

**Comment by jbogard** [comment]
Does the entire solution not load? I can load in VS2013, just the core project doesn't load. the AutoMapper.Net4 one does.

Anyway, it could be but I wanted to keep it there for posterity.

**Comment by TylerCarlson1** [comment]
Oh I can load it too as well.  I just assumed if you couldn't load Core then .Net4 wouldn't have the core base code.  I see now I can get to the code in VS2013.

**Comment by jbogard** [comment]
Yep, that's how I got the unit tests to work, since it's a regular .csproj. It's just readonly though, you can't add new files to the .Net4 one (it totally screws it up)

**Comment by TylerCarlson1** [comment]
I can't seem to run it, but I can see the code.  I have to do the tests in VS2015 and use MS Test.  I miss Resharper Test Runner so much.

**Comment by jbogard** [comment]
I think this one is good?

**Comment by TylerCarlson1** [comment]
I do too.
Original problem was solved with #727 
Everything after that all this was just discussion about the naming of functions, which isn't going to happen.
I say it's closed.
